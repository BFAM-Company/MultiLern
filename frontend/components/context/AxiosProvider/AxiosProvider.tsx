import React, { ReactNode, createContext, useContext } from "react";
import axios, { AxiosInstance } from "axios";
import { AuthContext } from "../AuthContext/AuthContext";
import createAuthRefreshInterceptor from "axios-auth-refresh";
import AsyncStorage from "@react-native-async-storage/async-storage";
// import * as Keychain from "react-native-keychain";

interface AuthState {
  accessToken: string | null;
  refreshToken: string | null;
}

interface AxiosContextProps {
  authAxios: AxiosInstance;
  publicAxios: AxiosInstance
}

const AxiosContext = createContext<AxiosContextProps>({
  authAxios: axios.create({
    baseURL: "http://localhost:3001/auth",
  }),
  publicAxios: axios.create({
    baseURL: "http://localhost:3001/",
  })
});

const { Provider } = AxiosContext;

const AxiosProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const authContext = useContext(AuthContext);

  const authAxios = axios.create({
    baseURL: "http://localhost:3001/auth",
  });

  const publicAxios = axios.create({
    baseURL: "http://localhost:3001/",
  });

  authAxios.interceptors.request.use(
    (config) => {
      if (!config.headers.Authorization) {
        config.headers.Authorization = `Bearer ${authContext?.authState.accessToken}`;
      }

      return config;
    },
    (error) => {
      return Promise.reject(error);
    }
  );

  const refreshAuthLogic = (failedRequest: any) => {
    const data = {
      refreshToken: authContext?.authState.refreshToken,
    };

    const options = {
      method: "POST",
      data,
      url: "http://localhost:3001/auth/refresh",
    };

    return axios(options)
      .then(async (tokenRefreshResponse: any) => {
        failedRequest.response.config.headers.Authorization =
          "Bearer " + tokenRefreshResponse.data;

        await AsyncStorage.setItem('auth',JSON.stringify({
            accessToken: tokenRefreshResponse.data,
            refreshToken: authContext?.authState.refreshToken,
        }))

        authContext?.setAuthState({
          ...authContext.authState,
          accessToken: tokenRefreshResponse.data,
        });
        
        return Promise.resolve();
      })
      .catch(async (e: any) => {
        authContext?.setAuthState({
          accessToken: null,
          refreshToken: null,
          authenticated: false
        });
        await AsyncStorage.clear()
      });
  };

  createAuthRefreshInterceptor(authAxios, refreshAuthLogic, {});

  return (
    <Provider
      value={{
        authAxios,
        publicAxios,
      }}
    >
      {children}
    </Provider>
  );
};

export { AxiosContext, AxiosProvider };
