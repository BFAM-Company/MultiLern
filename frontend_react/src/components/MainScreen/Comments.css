.comments-container {
  width: 100%;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: flex-start;
  padding: 20px;
  background-color: rgb(230, 230, 230);
  border-top-left-radius: 20px;
  border-top-right-radius: 20px;
  margin-top: 50px;
}

.bold-text {
  font-size: 20px;
  font-weight: 900;
  color: rgb(45, 45, 55);
}
