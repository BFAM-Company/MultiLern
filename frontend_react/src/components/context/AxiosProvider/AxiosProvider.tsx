import React, { ReactNode, createContext, useContext } from "react";
import axios, { AxiosInstance } from "axios";
import { AuthContext } from "../AuthContext/AuthContext";
import createAuthRefreshInterceptor from "axios-auth-refresh";

interface AuthState {
  accessToken: string | null;
  refreshToken: string | null;
}

interface AxiosContextProps {
  authAxios: AxiosInstance;
  publicAxios: AxiosInstance;
}

// Kontekst Axios
const AxiosContext = createContext<AxiosContextProps>({
  authAxios: axios.create({
    // baseURL: `${process.env.REACT_APP_API_URL}/auth`,
    baseURL: `${import.meta.env.VITE_API_URL}/auth`,
  }),
  publicAxios: axios.create({
    // baseURL: `${process.env.REACT_APP_API_URL}`,
    baseURL: `${import.meta.env.VITE_API_URL}`,
  }),
});

const { Provider } = AxiosContext;

const AxiosProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const authContext = useContext(AuthContext);

  const authAxios = axios.create({
    // baseURL: `${process.env.REACT_APP_API_URL}/auth`,
    baseURL: `${import.meta.env.VITE_API_URL}/auth`,
  });

  const publicAxios = axios.create({
    // baseURL: `${process.env.REACT_APP_API_URL}`,
    baseURL: `${import.meta.env.VITE_API_URL}`,
  });

  // Interceptor dodający nagłówki autoryzacji
  authAxios.interceptors.request.use(
    (config) => {
      if (authContext?.authState.accessToken) {
        config.headers.Authorization = `Bearer ${authContext.authState.accessToken}`;
      }
      return config;
    },
    (error) => Promise.reject(error)
  );

  // Funkcja obsługująca odświeżanie tokenu
  const refreshAuthLogic = (failedRequest: any) => {
    const data = {
      refreshToken: authContext?.authState.refreshToken,
    };

    return axios
      // .post(`${process.env.REACT_APP_API_URL}/auth/refresh`, data)
      .post(`${import.meta.env.VITE_API_URL}/auth/refresh`, data)
      .then((tokenRefreshResponse) => {
        const newAccessToken = tokenRefreshResponse.data;

        // Ustawienie nowego tokenu w localStorage
        localStorage.setItem(
          "auth",
          JSON.stringify({
            accessToken: newAccessToken,
            refreshToken: authContext?.authState.refreshToken,
          })
        );

        // Aktualizacja stanu kontekstu
        authContext?.setAuthState((prevState) => ({
          ...prevState,
          accessToken: newAccessToken,
        }));

        // Ponowienie oryginalnego żądania
        failedRequest.response.config.headers.Authorization = `Bearer ${newAccessToken}`;
        return Promise.resolve();
      })
      .catch(() => {
        // Wylogowanie w przypadku niepowodzenia
        authContext?.setAuthState({
          accessToken: null,
          refreshToken: null,
          authenticated: false,
        });
        localStorage.clear();
      });
  };

  // Dodanie interceptora obsługującego odświeżanie tokenu
  createAuthRefreshInterceptor(authAxios, refreshAuthLogic);

  return (
    <Provider value={{ authAxios, publicAxios }}>
      {children}
    </Provider>
  );
};

export { AxiosContext, AxiosProvider };
