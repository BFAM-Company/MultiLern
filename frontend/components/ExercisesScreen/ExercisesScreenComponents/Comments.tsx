import React, { useContext, useEffect, useState } from 'react';
import { View, Text, StyleSheet, Image } from 'react-native';
import SearchBar from '../../SearchBar/SearchBar';
import { AxiosContext } from '../../context/AxiosProvider';
import SingleComment from './SingleComment';
import { ActivityIndicator } from 'react-native-paper';
import CommentButton from './CommentButton';
      

interface CommentsProps{
    id: number
}

function Comments({id}: CommentsProps) {
    const {publicAxios, authAxios} = useContext(AxiosContext);
    const [comments, setComments] = useState<any[]>([])
    const [loading, setLoading] = useState<boolean>(false);


    useEffect(()=>{
        fetchPosts()
        }, [id])

    const fetchPosts = async() =>{
        setLoading(true)
        const result = await publicAxios.get(`/posts/postcomments/${id}`)
        if(result){
            //console.log(result.data)
            setLoading(false)
            setComments(result.data)
        }
    }

    function formatDate(isoDateString: string) {
        const months = [
            "styczeń", "luty", "marzec", "kwiecień", "maj", "czerwiec",
            "lipiec", "sierpień", "wrzesień", "październik", "listopad", "grudzień"
        ];
    
        const date = new Date(isoDateString);
        const day = date.getDate();
        const monthIndex = date.getMonth();
        const year = date.getFullYear();
    
        const formattedDate = `${day} ${months[monthIndex]} ${year}`;
        return formattedDate;
    }

    const calcRating = (exercise: any) =>{
        let rating = 0
        let counter = 0;
        //console.log(exercise.posts_reviews)
        if (exercise.posts_reviews && Array.isArray(exercise.posts_reviews)) {
          for(let i = 0; i < exercise.posts_reviews.length; i++){
            counter++
            rating += exercise.posts_reviews[i].reviews.rate
          }
          if (counter > 0) {
              return rating / counter;
          } else {
              return 0;
          }
        } else {
            return 0;
        }
      }
  return (
    <View
        style={styles.container}
    >
        <Text
            style={styles.boldText}
        >Odpowiedzi</Text>
        <View>
        {loading ? (
              <ActivityIndicator color="gray" style={{ margin: 15 }} />
            ) : null}
            {comments.map(comment => {
              const rating = calcRating(comment)
              const isoDateString = comment.date
              const formattedDate = formatDate(isoDateString);
              return(
                <SingleComment  
                  key={comment.id} 
                  user_data={comment.users_posts}
                  id={comment.id}
                  title={comment.title} 
                  description={comment.content} 
                  rate={rating} 
                  date={formattedDate} 
                  posts_images={comment.posts_images}/>)
            })}
            {
                !comments.length?(<Text style={{fontSize:18, color:'gray', fontWeight:'900', margin:10}}>Nic tu nie ma. Bądź pierwszy, który skomentuje!</Text>):null
            }

            <CommentButton id={id} handleRefresh={fetchPosts}/>
        </View>
    </View>
  );
}

const styles = StyleSheet.create({
    container:{
        width:'100%',
        display:'flex',
        flexDirection:'column',
        justifyContent:'flex-start',
        alignItems:'flex-start',
        padding:20,
        backgroundColor:'rgb(230,230,230)',
        borderTopLeftRadius:20,
        borderTopRightRadius:20,
        marginTop:50,
    },
    boldText:{
        fontSize:20,
        fontWeight:'900',
        color:'rgb(45,45,55)',
    },
})


export default Comments;