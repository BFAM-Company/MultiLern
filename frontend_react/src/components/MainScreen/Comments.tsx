import React, { useContext, useEffect, useState } from 'react';
import './Comments.css';
import { Spinner } from 'react-bootstrap';
import { AxiosContext } from '../context/AxiosProvider/AxiosProvider';
import SingleComment from './SingleComment';
import CommentButton from './CommentButton';

interface CommentsProps {
  id: number;
}

const Comments: React.FC<CommentsProps> = ({ id }) => {
  const { publicAxios } = useContext(AxiosContext);
  const [comments, setComments] = useState<any[]>([]);
  const [loading, setLoading] = useState<boolean>(false);

  useEffect(() => {
    fetchPosts();
  }, [id]);

  const fetchPosts = async () => {
    setLoading(true);
    try {
      const result = await publicAxios.get(`/posts/postcomments/${id}`);
      if (result) {
        setLoading(false);
        setComments(result.data);
      }
    } catch (error) {
      console.error('Error fetching posts:', error);
      setLoading(false);
    }
  };

  const formatDate = (isoDateString: string) => {
    const months = [
      'styczeń', 'luty', 'marzec', 'kwiecień', 'maj', 'czerwiec',
      'lipiec', 'sierpień', 'wrzesień', 'październik', 'listopad', 'grudzień'
    ];

    const date = new Date(isoDateString);
    const day = date.getDate();
    const monthIndex = date.getMonth();
    const year = date.getFullYear();

    return `${day} ${months[monthIndex]} ${year}`;
  };

  const calcRating = (exercise: any) => {
    let rating = 0;
    let counter = 0;

    if (exercise.posts_reviews && Array.isArray(exercise.posts_reviews)) {
      for (let i = 0; i < exercise.posts_reviews.length; i++) {
        counter++;
        rating += exercise.posts_reviews[i].reviews.rate;
      }
      return counter > 0 ? rating / counter : 0;
    } else {
      return 0;
    }
  };

  return (
    <div className="comments-container">
      <h2 className="bold-text">Odpowiedzi</h2>
      <div>
        {loading ? (
          <Spinner color="gray" style={{ margin: 15 }} />
        ) : null}

        {comments.length > 0 ? (
          comments.map((comment) => {
            const rating = calcRating(comment);
            const isoDateString = comment.date;
            const formattedDate = formatDate(isoDateString);
            return (
              <SingleComment
                key={comment.id}
                user_data={comment.users_posts}
                id={comment.id}
                title={comment.title}
                description={comment.content}
                rate={rating}
                date={formattedDate}
                posts_images={comment.posts_images}
              />
            );
          })
        ) : (
          <p style={{ fontSize: 18, color: 'gray', fontWeight: '900', margin: 10 }}>
            Nic tu nie ma. Bądź pierwszy, który skomentuje!
          </p>
        )}

        <CommentButton id={id} handleRefresh={fetchPosts} />
      </div>
    </div>
  );
};

export default Comments;
