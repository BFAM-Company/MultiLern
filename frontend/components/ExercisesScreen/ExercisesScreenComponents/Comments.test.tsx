import React from "react";
import { render, waitFor } from "@testing-library/react-native";
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import Comments from "./Comments";
import { AxiosContext } from "../../context/AxiosProvider/AxiosProvider";
import { ActivityIndicator } from 'react-native-paper';

jest.mock("@react-native-async-storage/async-storage", () => require("../../../__mocks__/mock-async-storage"));

describe('Comments', () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    it('renders correctly', () => {
        const { toJSON } = render(<Comments id={2} />);
        expect(toJSON()).toMatchSnapshot();
    });

    it('fetches posts and updates state', async () => {
        const commentsData = [
            { id: 1, title: "Comment 1", content: "This is a comment", date: "2023-05-29T00:00:00Z", users_posts: [{ users: { avatar: null } }], posts_images: [], posts_reviews: [{ reviews: { rate: 4 } }, { reviews: { rate: 5 } }] },
            { id: 2, title: "Comment 2", content: "This is another comment", date: "2023-05-28T00:00:00Z", users_posts: [{ users: { avatar: null } }], posts_images: [], posts_reviews: [{ reviews: { rate: 3 } }] }
        ];

        const mockPublicAxios = axios.create()
        mockPublicAxios.get = jest.fn().mockResolvedValue({data: commentsData})

        const { getByText, queryByText } = render(
            <AxiosContext.Provider value={{ publicAxios: mockPublicAxios, authAxios: mockPublicAxios }}>
                <Comments id={0} />
            </AxiosContext.Provider>
        );

        expect(queryByText('Nic tu nie ma. Bądź pierwszy, który skomentuje!')).toBeTruthy();

        await waitFor(() => {
            expect(queryByText('Nic tu nie ma. Bądź pierwszy, który skomentuje!')).toBeFalsy();
            commentsData.forEach(comment => {
                expect(getByText(comment.title)).toBeTruthy();
                expect(getByText(comment.content)).toBeTruthy();
                // Check formatted date
                const formattedDate = new Date(comment.date).toLocaleDateString('pl-PL', {
                    day: 'numeric',
                    month: 'short',
                    year: 'numeric'
                });
                expect(getByText(formattedDate)).toBeTruthy();
            });
        });
    });

    it('shows ActivityIndicator while loading', async () => {
        const mockPublicAxios = axios.create()
        mockPublicAxios.get = jest.fn().mockResolvedValue({data: []})
        const { getByTestId } = render(
            <AxiosContext.Provider value={{ publicAxios: mockPublicAxios, authAxios: mockPublicAxios }}>
                <Comments id={0} />
            </AxiosContext.Provider>
        );

        await waitFor(() => {
            expect(getByTestId('activity-indicator')).toBeTruthy();
        });
    });

    it('renders the comment button', () => {
        const mockPublicAxios = axios.create()
        mockPublicAxios.get = jest.fn().mockResolvedValue({data: []})
        const { getByText } = render(
            <AxiosContext.Provider value={{ publicAxios: mockPublicAxios, authAxios: mockPublicAxios }}>
                <Comments id={0} />
            </AxiosContext.Provider>
        );

        expect(getByText('Odpowiedz')).toBeTruthy();
    });
});
