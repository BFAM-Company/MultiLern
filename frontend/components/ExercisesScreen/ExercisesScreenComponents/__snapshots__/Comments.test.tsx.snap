// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "flex-start",
      "backgroundColor": "rgb(230,230,230)",
      "borderTopLeftRadius": 20,
      "borderTopRightRadius": 20,
      "display": "flex",
      "flexDirection": "column",
      "justifyContent": "flex-start",
      "marginTop": 50,
      "padding": 20,
      "width": "100%",
    }
  }
>
  <Text
    style={
      {
        "color": "rgb(45,45,55)",
        "fontSize": 20,
        "fontWeight": "900",
      }
    }
  >
    Odpowiedzi
  </Text>
  <View>
    <View
      accessibilityRole="progressbar"
      accessibilityState={
        {
          "busy": true,
        }
      }
      accessible={true}
      style={
        [
          {
            "alignItems": "center",
            "justifyContent": "center",
          },
          {
            "margin": 15,
          },
        ]
      }
      testID="activity-indicator"
    >
      <View
        collapsable={false}
        style={
          {
            "height": 24,
            "opacity": 1,
            "width": 24,
          }
        }
      >
        <View
          collapsable={false}
          style={
            {
              "alignItems": "center",
              "bottom": 0,
              "justifyContent": "center",
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              {
                "height": 24,
                "transform": [
                  {
                    "rotate": "45deg",
                  },
                ],
                "width": 24,
              }
            }
          >
            <View
              collapsable={false}
              style={
                {
                  "height": 12,
                  "overflow": "hidden",
                  "width": 24,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  {
                    "height": 24,
                    "transform": [
                      {
                        "translateY": 0,
                      },
                      {
                        "rotate": "-165deg",
                      },
                    ],
                    "width": 24,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    {
                      "height": 12,
                      "overflow": "hidden",
                      "width": 24,
                    }
                  }
                >
                  <View
                    collapsable={false}
                    style={
                      {
                        "borderColor": "gray",
                        "borderRadius": 12,
                        "borderWidth": 2.4,
                        "height": 24,
                        "width": 24,
                      }
                    }
                  />
                </View>
              </View>
            </View>
          </View>
        </View>
        <View
          collapsable={false}
          style={
            {
              "alignItems": "center",
              "bottom": 0,
              "justifyContent": "center",
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              {
                "height": 24,
                "transform": [
                  {
                    "rotate": "45deg",
                  },
                ],
                "width": 24,
              }
            }
          >
            <View
              collapsable={false}
              style={
                {
                  "height": 12,
                  "overflow": "hidden",
                  "top": 12,
                  "width": 24,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  {
                    "height": 24,
                    "transform": [
                      {
                        "translateY": -12,
                      },
                      {
                        "rotate": "345deg",
                      },
                    ],
                    "width": 24,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    {
                      "height": 12,
                      "overflow": "hidden",
                      "width": 24,
                    }
                  }
                >
                  <View
                    collapsable={false}
                    style={
                      {
                        "borderColor": "gray",
                        "borderRadius": 12,
                        "borderWidth": 2.4,
                        "height": 24,
                        "width": 24,
                      }
                    }
                  />
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
    <Text
      style={
        {
          "color": "gray",
          "fontSize": 18,
          "fontWeight": "900",
          "margin": 10,
        }
      }
    >
      Nic tu nie ma. Bądź pierwszy, który skomentuje!
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "borderRadius": 20,
          "height": 55,
          "margin": 20,
          "opacity": 1,
          "width": 300,
        }
      }
    >
      <ViewManagerAdapter_ExpoLinearGradient
        colors={
          [
            4281150775,
            4281150775,
            4284769390,
          ]
        }
        style={
          {
            "alignItems": "center",
            "borderRadius": 20,
            "display": "flex",
            "height": "100%",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "color": "white",
              "fontSize": 22,
              "fontWeight": "900",
            }
          }
        >
          Odpowiedz
        </Text>
      </ViewManagerAdapter_ExpoLinearGradient>
    </View>
  </View>
</View>
`;
